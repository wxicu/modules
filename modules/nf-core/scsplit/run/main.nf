process SCSPLIT_RUN {
    tag "$meta.id"
    label 'process_medium'

    conda "${moduleDir}/environment.yml"
    container 'wxicu/scsplit:1.0.8'

    input:
    tuple val(meta), path(ref), path(alt), val(num), path(vcf)

    output:
    tuple val(meta), path('*_result.csv')       , emit: result
    tuple val(meta), path('*_dist_variants.txt'), emit: dist_variants
    tuple val(meta), path('*_dist_matrix.csv')  , emit: dist_matrix
    tuple val(meta), path('*_PA_matrix.csv')    , emit: alt_allele
    tuple val(meta), path('*_P_s_c.csv')        , emit: probability
    tuple val(meta), path('*_scSplit.log')      , emit: log
    path 'versions.yml'                         , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def vcf_input = vcf ? "-v ${vcf}" : ''
    def VERSION = '1.0.8' // WARN: Version information not provided by tool on CLI. Please update version string below when bumping container versions.

    """
    if [ -z "${workflow.containerEngine}" ];
        then scsplit_path="scSplit";
    else
        scsplit_path="python \$(python -c 'import site; print("".join(site.getsitepackages()))')/scSplit/scSplit";
    fi
    \$scsplit_path run \\
        -r $ref \\
        -a $alt \\
        -n $num \\
        $vcf_input \\
        -o . \\
        $args

    mv scSplit_result.csv ${prefix}_result.csv
    mv scSplit_dist_variants.txt ${prefix}_dist_variants.txt
    mv scSplit_dist_matrix.csv ${prefix}_dist_matrix.csv
    mv scSplit_PA_matrix.csv ${prefix}_PA_matrix.csv
    mv scSplit_P_s_c.csv ${prefix}_P_s_c.csv
    mv scSplit.log ${prefix}_scSplit.log

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scsplit: $VERSION
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def VERSION = '1.0.8' // WARN: Version information not provided by tool on CLI. Please update version string below when bumping container versions.

    """
    touch ${prefix}_result.csv
    touch ${prefix}_dist_variants.txt
    touch ${prefix}_dist_matrix.csv
    touch ${prefix}_PA_matrix.csv
    touch ${prefix}_P_s_c.csv
    touch ${prefix}_scSplit.log

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scsplit: $VERSION
    END_VERSIONS
    """
}
