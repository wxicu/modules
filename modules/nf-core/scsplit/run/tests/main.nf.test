// nf-core modules test scsplit/run
nextflow_process {

    name "Test Process SCSPLIT_RUN"
    script "../main.nf"
    process "SCSPLIT_RUN"

    tag "modules"
    tag "modules_nfcore"
    tag "scsplit"
    tag "scsplit/run"
    tag "samtools/index"
    tag "scsplit/count"

    test("demultiplexing - bam") {

        setup {
            run("SAMTOOLS_INDEX") {
                script "../../../samtools/index/main.nf"
                process {
                    """
                    input[0] =[
                            [ id:'test'],
                            file(params.modules_testdata_base_path + '/genomics/homo_sapiens/demultiplexing/chr21.bam', checkIfExists: true)
                        ]

                    """
                }
            }

            run("SCSPLIT_COUNT") {
                script "../../count/main.nf"
                process {
                    """
                    bam_channel = Channel.of([
                                    [ id:'test'],
                                    file(params.modules_testdata_base_path + '/genomics/homo_sapiens/demultiplexing/chr21.bam', checkIfExists: true)
                                    ])
                    bai_channel = SAMTOOLS_INDEX.out.bai
                    vcf_barcode_channel = Channel.of([[ id:'test'],
                                                    file(/home/icb/xichen.wu/modules/modules/test.vcf),                                           
                                                    file(params.modules_testdata_base_path + '/genomics/homo_sapiens/demultiplexing/barcodes.tsv', checkIfExists: true)
                                                  ])

                    input[0] = bam_channel.join(bai_channel).join(vcf_barcode_channel)

                    """
                }
            }

        }
  
        when {
            process {
                """ 
                ref_channel = SCSPLIT_COUNT.out.ref_filtered
                alt_channel = SCSPLIT_COUNT.out.alt_filtered            
                num_channel = Channel.of([[id:'test'], 2])
                vcf_channel = Channel.of([[id:'test'], []])
                input[0] = ref_channel.join(alt_channel).join(num_channel).join(vcf_channel)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("demultiplexing - bam - stub") {

        options "-stub"

        setup {
            run("SAMTOOLS_INDEX") {
                script "../../../samtools/index/main.nf"
                process {
                    """
                    input[0] =[
                            [ id:'test'],
                            file(params.modules_testdata_base_path + '/genomics/homo_sapiens/demultiplexing/chr21.bam', checkIfExists: true)
                        ]

                    """
                }
            }

            run("SCSPLIT_COUNT") {
                script "../../count/main.nf"
                process {
                    """
                    bam_channel = Channel.of([
                                    [ id:'test'],
                                    file(params.modules_testdata_base_path + '/genomics/homo_sapiens/demultiplexing/chr21.bam', checkIfExists: true)
                                    ])
                    bai_channel = SAMTOOLS_INDEX.out.bai
                    vcf_barcode_channel = Channel.of([
                                            [ id:'test'],
                                            file(params.modules_testdata_base_path + '/genomics/homo_sapiens/demultiplexing/donor_genotype_chr21.vcf', checkIfExists: true),
                                            file(params.modules_testdata_base_path + '/genomics/homo_sapiens/demultiplexing/barcodes.tsv', checkIfExists: true),
                                            ])

                    input[0] = bam_channel.join(bai_channel).join(vcf_barcode_channel)

                    """
                }
            }

        }

        when {
            process {
                """                
                ref_channel = SCSPLIT_COUNT.out.ref_filtered
                alt_channel = SCSPLIT_COUNT.out.alt_filtered            
                num_channel = Channel.of([[id:'test'], 2])
                vcf_channel = Channel.of([[id:'test'], []])
                input[0] = ref_channel.join(alt_channel).join(num_channel).join(vcf_channel)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}
