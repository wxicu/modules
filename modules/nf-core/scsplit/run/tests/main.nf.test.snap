{
    "demultiplexing - bam": {
        "content": [
            [
                "versions.yml:md5,c8a05d1464c5632b94e87b2cc7c37c41"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-11T18:39:09.084922"
    },
    "demultiplexing - bam - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_result.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test_dist_variants.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test_dist_matrix.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test"
                        },
                        "test_PA_matrix.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test"
                        },
                        "test_P_s_c.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test"
                        },
                        "test_scSplit.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "6": [
                    "versions.yml:md5,c8a05d1464c5632b94e87b2cc7c37c41"
                ],
                "alt_allele": [
                    [
                        {
                            "id": "test"
                        },
                        "test_PA_matrix.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "dist_matrix": [
                    [
                        {
                            "id": "test"
                        },
                        "test_dist_matrix.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "dist_variants": [
                    [
                        {
                            "id": "test"
                        },
                        "test_dist_variants.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "log": [
                    [
                        {
                            "id": "test"
                        },
                        "test_scSplit.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "probability": [
                    [
                        {
                            "id": "test"
                        },
                        "test_P_s_c.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "result": [
                    [
                        {
                            "id": "test"
                        },
                        "test_result.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,c8a05d1464c5632b94e87b2cc7c37c41"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-11T18:39:50.548894"
    }
}